src/System.Text.Json/src/System/Text/Json/Serialization/ClassMaterializer.cs

```diff
@@ -14,25 +14,29 @@ internal abstract class ClassMaterializer
     {
         public abstract JsonClassInfo.ConstructorDelegate CreateConstructor(Type classType);
 
-        public abstract object ImmutableCreateRange(Type constructingType, Type elementType);
+        public abstract object ImmutableCreateRange(Type constructingType, Type elementType, bool constructingTypeIsDict);
 
-        protected MethodInfo ImmutableCreateRangeMethod(Type constructingType, Type elementType)
+        protected MethodInfo ImmutableCreateRangeMethod(Type constructingType, Type elementType, bool constructingTypeIsDict)
         {
             MethodInfo[] constructingTypeMethods = constructingType.GetMethods();
-            MethodInfo createRange = null;
 
             foreach (MethodInfo method in constructingTypeMethods)
             {
                 if (method.Name == "CreateRange" && method.GetParameters().Length == 1)
                 {
-                    createRange = method;
-                    break;
+                    if (constructingTypeIsDict)
+                    {
+                        return method.MakeGenericMethod(typeof(string), elementType);
+                    }
+                    else
+                    {
+                        return method.MakeGenericMethod(elementType);
+                    }
                 }
             }
 
-            Debug.Assert(createRange != null);
-
-            return createRange.MakeGenericMethod(elementType);
+            Debug.Fail("Could not create the appropriate CreateRange method.");
```

Comment:
Is continuing silently here (in release) the right thing to do? What will be the customer result - a partial deserialization, without any notification?

Reply:
I've modified this so it's clear we might return `null` here, then throw an exception further down if the value returned is `null`.